packages/manager/src/features/linodes/LinodesDetail/LinodeNetworking/IPSharing.tsx

```diff
@@ -103,22 +88,77 @@ const IPSharingPanel: React.FC<CombinedProps> = (props) => {
     open,
     onClose,
     linodeSharedIPs,
+    availableRanges,
   } = props;
 
-  const { data, isLoading } = useAllLinodesQuery(
+  const availableRangesMap: AvailableRangesMap = formatAvailableRanges(
+    availableRanges
+  );
+
+  const resp = useAllLinodesQuery(
     { page_size: API_MAX_PAGE_SIZE },
     {
       region: linodeRegion,
     },
     open // Only run the query if the modal is open
   );
 
+  let isLoading = resp.isLoading;
+  const data = resp.data;
+
+  const getIPChoicesAndLabels = (linodeID: number, linodes: Linode[]) => {
+    const choiceLabels = linodes.reduce((previousValue, currentValue) => {
+      // Filter out the current Linode
+      if (currentValue.id === linodeID) {
+        return previousValue;
+      }
+
+      currentValue.ipv4.forEach((ip) => {
+        previousValue[ip] = currentValue.label;
+      });
+
+      if (flags.ipv6Sharing) {
+        availableRangesMap?.[currentValue.id]?.forEach((range: string) => {
+          previousValue[range] = currentValue.label;
+          updateipToLinodeID({
+            [range]: [...(ipToLinodeID?.[range] ?? []), currentValue.id],
+          });
+        });
+      }
+
+      return previousValue;
+    }, {});
+
+    linodeSharedIPs.forEach((range) => {
+      if (!choiceLabels.hasOwnProperty(range)) {
+        choiceLabels[range] = '';
+      }
+    });
+
+    return choiceLabels;
+  };
+
   const linodes = data ?? [];
 
-  const { ipChoices, ipChoiceLabels } = React.useMemo(
-    () => getIPChoicesAndLabels(linodeID, linodes),
-    [linodeID, linodes]
-  );
+  const [ipChoices, setipChoices] = React.useState({});
+  const [ipToLinodeID, setipToLinodeID] = React.useState({});
+
+  const updateipToLinodeID = (newData: Record<string, number[]>) => {
+    setipToLinodeID((previousState) => {
+      return { ...previousState, ...newData };
+    });
+  };
+
+  React.useEffect(() => {
+    // don't try anything until we've finished the request for the Linodes data
+    if (isLoading) {
+      return;
+    }
+    isLoading = true;
+    const ipChoices = getIPChoicesAndLabels(linodeID, linodes);
+    setipChoices(ipChoices);
+    isLoading = false;
+  }, [linodeID, data, availableRanges]);
```

Comment:
We can remove the `linodes` variable and React Query handles the loading states for us so this can be simplified to

```suggestion
  React.useEffect(() => {
    // don't try anything until we've finished the request for the Linodes data
    if (!linodesData) {
      return;
    }
    const ipChoices = getIPChoicesAndLabels(linodeID, linodesData);
    setipChoices(ipChoices);
  }, [linodeID, linodesData, availableRanges]);
```